### ISSUE TITLE
Include contributions from The Kitchin Research Group

### ISSUE THREAD

**Issue by suhail-singh** [emacs-ffi/emacs-ffi#1]
Now that `emacs-ffi` is licensed under GPLv3, I would like to include
contributions from @jkitchin 's research group.  As noted in [this blog post](https://kitchingroup.cheme.cmu.edu/blog/2017/07/13/An-Emacs-zeromq-library-using-an-ffi/),
a few changes were made to `define-ffi-library` and `define-ffi-function`.  It's
possible, that since then, additional changes have been made as well.

While the blog post is licensed under CC-BY-SA 4.0 (which is [one-way compatible
with GPLv3](https://creativecommons.org/2015/10/08/cc-by-sa-4-0-now-one-way-compatible-with-gplv3/)),
I'd like to figure out the best possible way to collaborate with @jkitchin and
others in his group.  Specifically, would it be best for me to adapt the code as
I see fit for inclusion in my fork, or would you prefer to make pull request(s)?

Thoughts?

**Comment by jkitchin** [comment]
wow! I am happy to see this resurface. I would be happy to work on this with you. I am tied up for about a week, but maybe we could set up a meeting next Thur or Fri afternoon where we could chat about how to move forward?

**Comment by suhail-singh** [comment]
> I would be happy to work on this with you.

Awesome!

> I am tied up for about a week, but maybe we could set up a meeting next Thur
> or Fri afternoon where we could chat about how to move forward?

Sounds good.  I'll have a little more flexibility Friday afternoon.  What would
be the best way to reach you?  Is the email at `andrew.cmu.edu` the preferred
one?

**Comment by jkitchin** [comment]
yes that email is fine.

**Comment by suhail-singh** [comment]
@jkitchin it was wonderful connecting with you the other day.  For your
awareness, I will be packaging `emacs-ffi` for Guix some time in Jan 2025.  It
would be great for the initial packaged version to include some of the
quality-of-life tweaks you shared in your blog (and others should you have
them).  Regardless, I look forward to incorporating your changes when it's
feasible for you to open the pull request.

**Comment by jkitchin** [comment]
I am trying to get a github actions to work for testing. As far as I can tell it builds fine, but it cannot run the tests because it can't load a library (that it apparently used to build the module). See https://github.com/jkitchin/emacs-ffi/actions/runs/12529057472/job/34944263191. I assume this is some weird library path issue, but I can't figure out how to get it to work. any ideas?

**Comment by jkitchin** [comment]
I made some changes and suggestions in https://github.com/emacs-ffi/emacs-ffi/pull/4. The commit history isn't beautiful or particularly clean; a lot was just trying to get GitHUB Actions to work. It still only sort of works. I don't anticipate having a lot more time to work on it past this.

**Comment by suhail-singh** [comment]
@jkitchin thank you for #4.

> a lot was just trying to get GitHUB Actions to work. It still only sort of
> works.

This was one of the things I wanted to get done, so thank you for working on
this.  Saves me effort not having to start from scratch.

> I don't anticipate having a lot more time to work on it past this.

All good.  I'll make tweaks as desired and get it merged in the next week or so.

**Comment by tarsius** [comment]
> I'll make tweaks as desired and get it merged in the next week or so.

Could you please let some time pass between these two steps so I get a chance to review?

**Comment by suhail-singh** [comment]
> Could you please let some time pass between these two steps so I get a chance
> to review?

The two steps being making the tweaks, and merging the changes?  If so, sure.
Out of curiosity, is your interest specific to the Github actions work, or you
mean in general?

**Comment by tarsius** [comment]
Yes, those steps. And I am interested in general, with a focus on the actions. 😁

**Comment by suhail-singh** [comment]
> And I am interested in general, with a focus on the actions.

Sure, happy to leave the PR open for a few days after I make changes.

That being said, for github actions specifically it may make more sense for you
to not wait for me (I have more experience with gitlab's CI than github's)
provided you have time to spare.

**Comment by suhail-singh** [comment]
@tarsius i'm inclined to accept the changeset in #4 with no or minimal (e.g.,
whitespace tweaks) changes.  I'll tackle the github action issue(s) separately
(and i'll open a separate pull request for that).

Please let me know if you have objections or need some time to review the
existing changeset.

**Comment by tarsius** [comment]
Please do not merge this without first turning it into a meaningful sequence of commits that can actually be reviewed.

**Comment by suhail-singh** [comment]
> Please do not merge this without first turning it into a meaningful sequence
> of commits that can actually be reviewed.

As in squash commits and revise commit message?  I agree that that would be
useful (at least to the point that git-blame gives something informative).  That
being said, I'm inclined to spend limited time on such cleanup, and it doesn't
seem John would have much time on that either.

Would you like to do the cleanup yourself, or would you be okay with the limited
cleanup that I do?  I prefer not to have long outstanding changesets, and while
I care about commits being broken into meaningful chunks, I will at times
compromise on that due to resource constraints.

**Comment by tarsius** [comment]
I've started working on it again. Please hold on a little longer.

**Comment by suhail-singh** [comment]
> Please hold on a little longer.

Okay, I can check back in a few days.  There are some useful conveniences in #4
that I'd like to get merged by next week, but we can discuss in a few days and
assess the situation.

**Comment by suhail-singh** [comment]
> I've started working on it again. Please hold on a little longer.

@tarsius any updates?  For my understanding, are you simply doing a cleanup of
the history (i.e., squashing commits etc.), or are you also attempting to
address the failure in tests noted by John in the GH actions?

**Comment by suhail-singh** [comment]
I have merged #4 into the `dev` branch.  I've done some minor cleanup of
`README.org` and it now replaces `README.md`.  Once I have tested the code
snippets, I'll merge `dev` into `master` and close this issue.

Improvements and fixes regarding the github action issue(s) will be tackled as
part of #5.
